core/src/main/scala/kafka/message/ByteBufferMessageSet.scala

```diff
@@ -226,48 +357,215 @@ class ByteBufferMessageSet(val buffer: ByteBuffer) extends MessageSet with Loggi
+
+      var maxTimestamp = Message.NoTimestamp
+      val expectedRelativeOffset = new AtomicLong(0)
+      val validatedMessages = new ListBuffer[Message]
+      this.internalIterator(isShallow = false).foreach(messageAndOffset => {
+        val message = messageAndOffset.message
+        validateMessageKey(message, compactedTopic)
+        if (message.magic > Message.MagicValue_V0 && magicValueToUse > Message.MagicValue_V0) {
+          // Re-compression situation 3
+          // Validate the timestamp
+          validateTimestamp(message, now, messageTimestampType, messageTimestampDiffMaxMs)
+          // Check if we need to overwrite offset
+          if (messageAndOffset.offset != expectedRelativeOffset.getAndIncrement)
+            requireReCompression = true
+          maxTimestamp = math.max(maxTimestamp, message.timestamp)
+        }
+
+        // Re-compression situation 4
+        if (message.magic != magicValueToUse)
+          requireReCompression = true
+
+        validatedMessages += message.toFormatVersion(magicValueToUse)
       })
-      new ByteBufferMessageSet(compressionCodec = targetCodec, offsetCounter = offsetCounter, messages = messages.toBuffer:_*)
+
+      if (requireReCompression) {
+        // Re-compression required.
+        val messageSetTimestampAssignor = (messages: Seq[Message]) => {
+          if (magicValueToUse == Message.MagicValue_V0)
+            Message.NoTimestamp
+          else if (magicValueToUse > Message.MagicValue_V0 && messageTimestampType == TimestampType.CreateTime)
+            maxTimestamp
+          else // Log append time
+            now
+        }
+
+        new ByteBufferMessageSet(compressionCodec = targetCodec,
+                                 offsetCounter = offsetCounter,
+                                 messageSetTimestampAssignor = messageSetTimestampAssignor,
+                                 timestampType = messageTimestampType,
+                                 messages = validatedMessages.toBuffer: _*)
+      } else {
+        // Do not do re-compression but simply update the offset, timestamp and attributes field of the wrapper message.
+        buffer.putLong(0, offsetCounter.addAndGet(validatedMessages.size) - 1)
+
+        if (magicValueToUse > Message.MagicValue_V0) {
+          var crcUpdateNeeded = true
+          val timestampOffset = MessageSet.LogOverhead + Message.TimestampOffset
+          val attributeOffset = MessageSet.LogOverhead + Message.AttributesOffset
+          val timestamp = buffer.getLong(timestampOffset)
+          val attributes = buffer.get(attributeOffset)
+          if (messageTimestampType == TimestampType.CreateTime) {
+            if (timestamp == maxTimestamp && TimestampType.getTimestampType(attributes) == TimestampType.CreateTime)
+              // We don't need to recompute crc if the timestamp is not updated.
+              crcUpdateNeeded = false
+            else {
+              // Set timestamp type and timestamp
```

Comment:
For migration case and for producers that is not setting the timestamp of outer message correctly somehow. It should be fine since we are not changing the actual timestamps of messages.